src/types/capsule.rs

```diff
@@ -198,27 +209,64 @@ pub trait PyCapsuleMethods<'py>: crate::sealed::Sealed {
+    unsafe fn reference<T>(&self) -> &T;
 
-    /// Gets the raw `c_void` pointer to the value in this capsule.
+    /// Obtains a reference dereferenced from the pointer of this capsule. This is inherently very
+    /// dangerous: it involves casting an untyped pointer to a specific type. Additionally, arbitrary
+    /// Python code can change the contents of the capsule, which may invalidate the reference.
+    ///
+    /// Returns an error if the `name` does not exactly match the name stored in the capsule, or if
+    /// the pointer stored in the capsule is null.
     ///
-    /// Returns null if this capsule is not valid.
+    /// # Safety
+    ///
+    /// - It must be known that the capsule pointer points to an item of type `T`.
+    /// - The reference should not be used after arbitrary Python code has run.
+    ///
+    /// It is recommended to use this reference only for short-lived operations without executing any
+    /// Python code. For long-lived operations, consider calling `.reference_checked()` each time the
+    /// data is needed.
+    unsafe fn reference_checked<T>(&self, name: Option<&CStr>) -> PyResult<&T>;
+
+    /// Gets the raw pointer stored in this capsule, without checking its name.
+    #[deprecated(since = "0.27.0", note = "use `pointer_checked()` instead")]
     fn pointer(&self) -> *mut c_void;
 
-    /// Checks if this is a valid capsule.
+    /// Gets the raw pointer stored in this capsule.
     ///
-    /// Returns true if the stored `pointer()` is non-null.
+    /// Returns an error if the capsule is not [valid][`PyCapsuleMethods::is_valid`] with the given `name`.
+    fn pointer_checked(&self, name: Option<&CStr>) -> PyResult<NonNull<c_void>>;
+
+    /// Checks if the capsule pointer is not null.
+    ///
+    /// This does not perform any check on the name of the capsule, which is the only mechanism
+    /// that Python provides to make sure that the pointer has the expected type. Prefer to use
+    /// [`is_valid_checked()`][Self::is_valid_checked()] instead.
+    #[deprecated(since = "0.27.0", note = "use `is_valid_checked()` instead")]
     fn is_valid(&self) -> bool;
 
-    /// Retrieves the name of this capsule, if set.
+    /// Checks that the capsule name matches `name` and that the pointer is not null.
+    fn is_valid_checked(&self, name: Option<&CStr>) -> bool;
+
+    /// Retrieves the name of this capsule. If there is no name, the pointer will be null.
     ///
     /// Returns an error if this capsule is not valid.
-    fn name(&self) -> PyResult<Option<&'py CStr>>;
+    ///
+    /// This method returns `*const c_char` instead of `&CStr` because it's possible for
+    /// arbitrary Python code to change the capsule name. Callers can use `NonNull::from_ptr()`
+    /// to get a `&CStr` if they want to, however they should beware the fact that the pointer
+    /// may become invalid after arbitrary Python code has run.
+    fn name(&self) -> PyResult<*const c_char>;
```

Comment:
I changed this because the `&CStr` had the same problematic lifetime as `.reference()`. Also `CStr::from_ptr` is linear complexity so this is potentially more efficient depending on what the user is doing with it.